src/meta-model.ts

```diff
@@ -0,0 +1,237 @@
+import {util} from "./util";
+const isInteger = require("is-integer");
+const indentString = require("indent-string");
+
+export function buildMetaModel(model: IMetaModel): MetaModel {
+	return new MetaModel(model);
+}
+
+export function buildSemantics(models: any[]) {
+	return new SemanticsNode(models);
+}
+
+/*
+ * The following type definitions _must_ match with the metamodel json.
+ */
+
+export const TYPE_KEY_NAME = "$sType";
+
+export interface IMetaModel {
+	[typeName: string]: IMetaType;
+}
+
+export interface IMetaType {
+	properties: { [propertyName: string]: IPropertyDescription };
+	description: any;
+	initialValue: Object;
+}
+
+export type IType = "string" | "Object" | "any";
+
+export interface IPropertyDescription {
+	type: IType;
+	stringMap?: boolean;
+	optional?: boolean;
+	ownType?: boolean;
+}
+
+/**
+ * The default instance, exported from this file,
+ * of this class encapsulates the meta model,
+ * i.e. a formal description of all sTypes.
+ */
+export class MetaModel {
+	private _model: IMetaModel;
+
+	constructor(model: IMetaModel) {
+		this._model = model;
+	}
+
+	sTypes(): string[] {
+		return Object.keys(this._model);
+	}
+
+	createOfSType(sType: string) {
+		if (this.sTypes().indexOf(sType) < 0) {
+			throw new Error(`Cannot create instance of: ${sType}`);
+		}
+
+		return {
+			[TYPE_KEY_NAME]: sType,
+			...this._model[sType].initialValue
+		};
+	}
+}
+
+/**
+ * @returns whether the given JSON value is a semantically-typed object.
+ */
+export function isSemanticsTyped(json: any) {
+	if (util.isObject(json)) {
+		const sTypeValue = json[TYPE_KEY_NAME];
+		return sTypeValue && util.isString(sTypeValue);
+	}
+	return false;
+}
+
+/**
+ * @returns the semantics type from the given object.
+ */
+export function	sType(object: Object) {
+	return object[TYPE_KEY_NAME];
+}
+
+export class PropertyNode {
```

Comment:
`SemanticsNode`, `PropertyNode` and `InterfaceNode` These classes are responsible for generating the meta model typings. Each have a `print()` method that does their own thing using different templates. So `SemanticsNode(json).print()` gives you the typings.